File: 0629 - K Inverse Pairs Array/solution.cpp
class Solution {
public:
    int kInversePairs(int n, int k) {
        const int MOD = 1e9+7;
        vector<vector<int>> dp(2, vector<int>(k+1));
        dp[0][0] = 1;
        for(int i = 2; i <= n; i++){
            dp[1][0] = 1;
            for(int j = 1; j <= k; j++){
                dp[1][j] = (dp[1][j-1]+dp[0][j])%MOD;
                if(j >= i) dp[1][j] = (dp[1][j]-dp[0][j-i]+MOD)%MOD;
            }
            swap(dp[0], dp[1]);
        }
        return dp[0][k];
    }
};

File: 0629 - K Inverse Pairs Array/solution.py
class Solution:
    def kInversePairs(self, n: int, k: int) -> int:
        dp1, dp2, MOD = [1]+[0]*k, [0]*(k+1), int(1e9+7)
        for i in range(2, n+1):
            dp2[0] = 1
            for j in range(1, k+1):
                dp2[j] = MOD-dp1[j-i] if j >= i else 0
                dp2[j] = (dp2[j]+dp2[j-1]+dp1[j])%MOD
            dp1, dp2 = dp2, dp1
        return dp1[k]%MOD